src/internal/modules/address-entry/controller.js

```diff
@@ -0,0 +1,98 @@
+const { ukPostcode, selectAddress, manualAddressEntry } = require('./forms');
+const forms = require('shared/lib/forms');
+const newAddress = require('../../lib/new-address');
+const helpers = require('./lib/helpers');
+const { omit } = require('lodash');
+
+const sessionForms = require('shared/lib/session-forms');
+
+const storeAddressAndRedirect = (request, h, address) => {
+  newAddress.set(request, address);
+  const redirectPath = helpers.getRedirectPath(request);
+  return h.redirect(redirectPath);
+};
+
+const getPostcode = (request, h) => {
+  helpers.saveReferenceData(request);
+  return h.view('nunjucks/address-entry/enter-uk-postcode', {
+    ...request.view,
+    ...helpers.getPageCaption(request),
+    pageTitle: 'Enter the UK postcode',
+    back: request.query.back,
+    form: sessionForms.get(request, ukPostcode.form(request))
+  });
+};
+
+const postPostcode = async (request, h) => {
+  const { postcode } = request.payload;
+  const form = forms.handleRequest(
+    ukPostcode.form(request, postcode),
+    request,
+    ukPostcode.schema
+  );
+
+  if (form.isValid) {
+    return h.redirect(helpers.getSelectAddressUrl(request));
+  }
+  const { redirectPath, back } = request.yar.get(helpers.SESSION_KEY);
+  return h.postRedirectGet(form, '/address-entry/postcode', { redirectPath, back });
+};
+
+const getSelectAddress = (request, h) => {
+  helpers.setAddressSearchResults(request);
```

Comment:
I think I'd be inclined to put the postcode in a hidden field rather than store the whole address search result set in the session.
This makes the POST handler stateless which is always desirable.
The POST handler would need to make another request to the water service with the postcode and then find the address by UPRN.
To save API requests to the address service, which cost, Redis backed cache could be implemented in the water service, and the UI would not need to worry.